src/function/table/system/attach_duckdb.cpp

```diff
@@ -0,0 +1,241 @@
+#include <stdint.h>
+#include "duckdb.hpp"
+#include "duckdb/main/query_result.hpp"
+#include "duckdb/parser/parser.hpp"
+#include "duckdb/parser/expression/cast_expression.hpp"
+#include "duckdb/parser/parsed_data/create_table_function_info.hpp"
+#include "duckdb/common/types/date.hpp"
+#include "duckdb/common/types/timestamp.hpp"
+
+#include <cmath>
+
+namespace duckdb {
+
+// this is private in query_result for some reason, I'd rather include it from there but for now I'll do this
+namespace __copied_from_query_result_hpp {
+class QueryResultIterator;
+class QueryResultRow {
+public:
+	explicit QueryResultRow(QueryResultIterator &iterator_p, idx_t row_idx) : iterator(iterator_p), row(0) {
+	}
+
+	QueryResultIterator &iterator;
+	idx_t row;
+
+	Value GetValue(idx_t col_idx) const;
+};
+//! The row-based query result iterator. Invoking the
+class QueryResultIterator {
+public:
+	explicit QueryResultIterator(QueryResult *result_p) : current_row(*this, 0), result(result_p), base_row(0) {
+		if (result) {
+			chunk = shared_ptr<DataChunk>(result->Fetch().release());
+			if (!chunk) {
+				result = nullptr;
+			}
+		}
+	}
+
+	QueryResultRow current_row;
+	shared_ptr<DataChunk> chunk;
+	QueryResult *result;
+	idx_t base_row;
+
+public:
+	void Next() {
+		if (!chunk) {
+			return;
+		}
+		current_row.row++;
+		if (current_row.row >= chunk->size()) {
+			base_row += chunk->size();
+			chunk = result->Fetch();
+			current_row.row = 0;
+			if (!chunk || chunk->size() == 0) {
+				// exhausted all rows
+				base_row = 0;
+				result = nullptr;
+				chunk.reset();
+			}
+		}
+	}
+
+	QueryResultIterator &operator++() {
+		Next();
+		return *this;
+	}
+	bool operator!=(const QueryResultIterator &other) const {
+		return result != other.result || base_row != other.base_row || current_row.row != other.current_row.row;
+	}
+	const QueryResultRow &operator*() const {
+		return current_row;
+	}
+};
+
+Value QueryResultRow::GetValue(idx_t col_idx) const {
+	return iterator.chunk->GetValue(col_idx, row);
+}
+} // namespace __copied_from_query_result_hpp
+using QueryResultIterator = __copied_from_query_result_hpp::QueryResultIterator;
+
+struct ExternalQueryFunctionData : public TableFunctionData {
+	ExternalQueryFunctionData() {
+	}
+
+	bool finished = false;
+	string file_name = "";
+	string table = "";
+	unique_ptr<DuckDB> database = nullptr;
+	unique_ptr<Connection> connection = nullptr;
+	unique_ptr<QueryResult> tablerelation = nullptr;
+	unique_ptr<QueryResultIterator> rowiterator = nullptr;
+};
+
+static unique_ptr<FunctionData> ExternalQueryBind(ClientContext &context, TableFunctionBindInput &input,
+                                                  vector<LogicalType> &return_types, vector<string> &names) {
+
+	auto result = make_unique<ExternalQueryFunctionData>();
+	result->file_name = input.inputs[0].GetValue<string>();
+	result->table = input.inputs[1].GetValue<string>();
+
+	result->database = make_unique<DuckDB>(result->file_name);
```

Comment:
This should probably open a read-only connection to avoid contention. Ideally the buffer manager and thread pool are also shared between the instances, but we can implement that as an enhancement later.